### PR TITLE
Adding Tests and Whatnot

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR adds basic tests, helpers, migrations (for tests) and fixtures. Some notes:
- I ended up swapping out AVA for mocha. Sorry for the fail there, but I was getting constrained by AVA in a few ways that was harshing my velocity (tight, tight timelines right now). Glad to discuss further with things to back up my reasoning.
- You'll note some non-implemented tests. Those are ones for which I kicked up some bugs (or design quirks) and wanted to re-address not within this branch.

Regarding bugs or confusions:
- One of the SQL queries was missing a comma and was an obvious and quick error to fix. So I did fix that one on this branch.
- I want to take a higher-level look at response codes. I'm getting 409s when I'd expect 400s in several cases. 500s in a few cases.
- Requests that violate DB-level uniqueness constraints are throwing (unhandled) exceptions.
- PATCH requests are failing because of schema validations (more fields are required in the request body than should be).
- Invalid DELETE requests still return 204s (maybe that's as should be but I wanted to look it up)

I'm happy to look at all of the above, but, one branch per thing :). And it looks like some of the `pg-promise` stuff needs to be refactored anyway (based on new features in the package)...

These are API-consumption tests only; not sure if you'd like to see tests at different granularities...let me know.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,27 +7,38 @@
     "migrate:up": "careen -A -e -c db/migrations/careen",
     "migrate:down": "careen -R -e -c db/migrations/careen",
     "migrate:create": "careen -C -c db/migrations/careen",
+    "pretest": "dropdb test-api-boilerplate || true && createdb test-api-boilerplate",
     "seed": "careen -c db/seeds/careen",
     "seed:up": "careen -A -e -c db/seeds/careen",
     "seed:down": "careen -R -e -c db/seeds/careen",
-    "seed:create": "careen -C -c db/seeds/careen"
+    "seed:create": "careen -C -c db/seeds/careen",
+    "test": "PGNAME=test-api-boilerplate mocha --compilers js:babel-register"
```
Comment (lyzadanger) [inline package.json]: PGNAME is a slightly awkward ENV to use to indicate TEST, BUT I'm taking cues from 12-factor not to make ENVs into fake "environment" combinations. So...I'm open to feedback on how to handle this!

File: test/helpers/migrate.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+export const up = execute.bind(null, '-A -e');
+
+/**
+ * Take DB back, migrate down
+ * -R revert migrations (down)
+ * (TODO: -e and -a in the same command?)
+ * TODO: How to derive ID of first migration?
+ */
+export const reset = function () {
+  execute('-R -a -e --to=20150823131441');
```
Comment (lyzadanger) [inline test/helpers/migrate.js]: Obviously, THIS IS WRONG, but I'm not sure how best to handle this given that I don't necessarily know the IDs for the migrations.

I'm not using this method yet...so.

Comment (tkellen) [reply test/helpers/migrate.js]: use `-n 10000` (that's number of migrations and it doesn't throw if you tell it too many).

File: package.json (hunk 1 above)
Comment (tkellen) [reply package.json]: You can remap multiple options in `config.js` if you want to, so like,

``` js
  db: {
    host: process.env.PGHOST,
    port: process.env.PGPORT,
    database: process.env.PGNAME || process.env.DB_NAME || NAME,
    user: process.env.PGUSER,
    password: process.env.PGPASS
  }
```

File: test/fixtures/state.json [hunk 3]
```diff
@@ -0,0 +1,52 @@
+[
+  { "name" : "ALABAMA", "abbr" : "AL" },
```
Comment (tkellen) [inline test/fixtures/state.json:2]: For explicitness, I would put the "id" field in here as well.

File: test/helpers/fixtures.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+export default function fixtures () {
+  // Array of promises representing table-insert batch transactions
+  const insertPromises = [];
+  for (const table in data) {
+    // Insert data for each table
+    insertPromises.push(db.tx(t => {
+      return t.batch(data[table].map(insert => {
+        const fields = Object.keys(insert);
```
Comment (tkellen) [inline test/helpers/fixtures.js:15]: this _should_ give a single multi-value insert statement:

``` js
const columns = Object.keys(data.state[0]);
return pgp.helpers(data.state, columns, 'state');
```

**Comment by tkellen** [comment]
Happy to talk error codes, btw. When I don't feel like digging into the RFCs, I follow this:


Hugely appreciating the tests, there are clearly a lot of holes in this patchwork template. As a heads up, I am out next week. I just added you as a collab on this repo. Feel free to land anything you'd like!

File: test/helpers/migrate.js (hunk 2 above)
Comment (tkellen) [reply test/helpers/migrate.js]: PS: if this cruft bugs you too much, the only other library I'd recommend is
https://github.com/db-migrate

...it does provide a nice API, but it doesn't support straight SQL without an intermediate JS file to call them for every migration. I'm not personally wild about that, but it's just exchanging one type of cruft for another, so, your call!
https://db-migrate.readthedocs.io/en/latest/Getting%20Started/usage/

Comment (lyzadanger) [reply test/helpers/migrate.js]: Sounds good for the future (thanks for all the info!) but in the interest of celerity I think I'll use your `-n` flag for the time being!

File: test/helpers/fixtures.js (hunk 4 above)
Comment (lyzadanger) [reply test/helpers/fixtures.js:15]: Huh. This one's going to take some unraveling. Right now, `src/services/db` has a single (default) export that is the resulting `db` object from munged config options. That is, I don't have access to the `pgp` library itself from that exported object (I don't think? `pg-promise` is new to me and I haven't gotten my sea legs yet). Here are the options that leap to mind:
- Add an additional, named, export to `db`. Pros: easy as pie. Cons: combining named and default exports. I don't know what our position is on that :).
- Export something else (e.g. the entire `pgp` object) from `db`. Yikes.
- Re-initialize `pgp` in places where it's needed. Also yikes (let's leave config munging in one place, eh?).

I guess given those options I'll probably head toward the multiple-exports-from-`db` one. I think we'll probably run into other places we need access to the `pgp` library itself...you think? Anyway, I'll check with `#code-review` to see if there is a backlash against this form of exporting and go that way if not.

Comment (tkellen) [reply test/helpers/fixtures.js:15]: Whoops, that was a typo on my part. You should have `db.helpers` right off the default export!

Comment (lyzadanger) [reply test/helpers/fixtures.js:15]: Alas, nope:

```
Database {
  connect: [Function],
  query: [Function],
  none: [Function],
  one: [Function],
  many: [Function],
  oneOrNone: [Function],
  manyOrNone: [Function],
  any: [Function],
  result: [Function],
  stream: [Function],
  func: [Function],
  proc: [Function],
  map: [Function],
  each: [Function],
  task: [Function],
  tx: [Function] }
```

Comment (tkellen) [reply test/helpers/fixtures.js:15]: Ah boo. Okay. I'd throw the `pgp` export in `src/base/pgp.js` and import it in `services/db`.

`src/services/` is meant to hold files related to connections with external services. It doesn't make sense (to my mind) to export the constructor/factory for connecting to one from there. However, `src/base` seems like the perfect location.

Comment (lyzadanger) [reply test/helpers/fixtures.js:15]: Makes sense! And thank you for responding to this thread even when on PTO! That's nuts. Go have fun!

**Comment by lyzadanger** [comment]
To keep this PR/branch from getting too rambly, I'd like to calve off some of the feedback here into separate PRs. Specifically:
- A branch/PR for `pg-promise` (DB) updates, including query helper use, creating `base/src/pgp`, and updating the way fixture population works to use a `pg-promise` helper. Additionally, I'm having some woes adding the explicit `id` column data to my fixture JSON as suggested and I don't want to spin my wheels solving it if I'm going to make changes to queries underneath it later (and potentially break it again). So I'd like to do that as part of this work as well.
- A branch/PR for addressing response codes.

Other feedback has been addressed, so I'm going to merge now (ask for forgiveness later ;) ) and start work on the follow-up stuff in new branches!

**Pull request merged**